app/Platform/Actions/UpdatePlayerPointsStatsBatchAction.php

```diff
@@ -0,0 +1,127 @@
+
+class UpdatePlayerPointsStatsBatchAction
+{
+    use CalculatesPlayerPointsStats;
+
+    /**
+     * Process a batch of users and update their points stats using bulk operations.
+     *
+     * @param array<int> $userIds
+     */
+    public function execute(array $userIds): void
+    {
+        if (empty($userIds)) {
+            return;
+        }
+
+        // Get users and handle untracked users.
+        $users = User::whereIn('id', $userIds)->get()->keyBy('id');
+
+        // Wipe stats for untracked users.
+        $untrackedUserIds = $users->where('Untracked', true)->pluck('id')->toArray();
+        if (!empty($untrackedUserIds)) {
+            PlayerStat::whereIn('user_id', $untrackedUserIds)->delete();
+        }
+
+        // Process tracked users.
+        $trackedUsers = $users->where('Untracked', false);
+        if ($trackedUsers->isEmpty()) {
+            return;
+        }
+
+        $trackedUserIds = $trackedUsers->pluck('ID')->toArray();
+
+        // Fetch all player achievements for tracked users in the time window.
+        $allAchievements = PlayerAchievement::whereIn('player_achievements.user_id', $trackedUserIds)
+            ->whereBetween('player_achievements.unlocked_at', [Carbon::now()->subDays(8), Carbon::now()])
+            ->join('Achievements', 'player_achievements.achievement_id', '=', 'Achievements.ID')
+            ->join('GameData', 'Achievements.GameID', '=', 'GameData.ID')
+            ->where('GameData.ConsoleID', '!=', System::Events)
+            ->select(
+                'player_achievements.user_id',
+                'player_achievements.unlocked_at',
+                'player_achievements.unlocked_hardcore_at',
+                'Achievements.Points as points',
+                'Achievements.TrueRatio as points_weighted'
+            )
+            ->get()
+            ->groupBy('user_id');
+
+        // Calculate stats for each user.
+        $statIntervals = $this->getStatIntervals();
+        $bulkStats = [];
+
+        foreach ($trackedUserIds as $userId) {
+            $userAchievements = $allAchievements->get($userId, collect());
+
+            // Separate achievements by type.
+            ['hardcore' => $hardcoreAchievements, 'softcore' => $softcoreAchievements] =
+                $this->separateAchievementsByType($userAchievements);
+
+            foreach ($statIntervals as $period => $interval) {
+                $hardcorePoints = $this->calculatePointsForInterval($hardcoreAchievements, $interval);
+                $softcorePoints = $this->calculatePointsForInterval($softcoreAchievements, $interval);
+
+                $periodStats = $this->buildStatsForPeriod($userId, $hardcorePoints, $softcorePoints, $period);
+
+                // Add timestamps for bulk insert.
+                foreach ($periodStats as $stat) {
+                    $bulkStats[] = array_merge($stat, [
+                        'created_at' => now(),
+                        'updated_at' => now(),
```

Comment:
Not sure how much overhead `now()` has, but consider calculating it once outside the outer loop.